service/history/task_executor.go

```diff
@@ -0,0 +1,203 @@
+			l.metricsHandler.Counter(metrics.StaleMutableStateCounter.Name()).Record(1)
+			wfCtx.Clear()
+			continue
+		}
+		return nil, err
+	}
+}
+
+// validateTaskVersion validates task version against the given stateVersion.
+// If namespace is not a global namepace, this always returns nil.
+// The returned error is always a [queues.TaskStateMismatchError].
+func (l *taskExecutor) validateTaskVersion(
+	ns *namespace.Namespace,
+	stateVersion int64,
+	task tasks.Task,
+) error {
+	if !l.shardContext.GetClusterMetadata().IsGlobalNamespaceEnabled() {
+		return nil
+	}
+
+	// the first return value is whether this task is valid for further processing
+	if !ns.IsGlobalNamespace() {
+		l.logger.Debug("NamespaceID is not global, task version check pass", tag.WorkflowNamespaceID(ns.ID().String()), tag.Task(task))
+		return nil
+	}
+	if stateVersion < task.GetVersion() {
+		l.logger.Debug("NamespaceID is global, state version < task version", tag.WorkflowNamespaceID(ns.ID().String()), tag.Task(task), tag.TaskVersion(task.GetVersion()))
+		return queues.NewTaskStateMismatchError("state version < task version", true)
+	}
+	if stateVersion > task.GetVersion() {
+		l.logger.Debug("NamespaceID is global, state version > task version", tag.WorkflowNamespaceID(ns.ID().String()), tag.Task(task), tag.TaskVersion(task.GetVersion()))
+		return queues.NewTaskStateMismatchError("state version > task version", false)
+	}
+	l.logger.Debug("NamespaceID is global, task version == target version", tag.WorkflowNamespaceID(ns.ID().String()), tag.Task(task), tag.TaskVersion(task.GetVersion()))
+	return nil
+}
+
+// validateCallbackTask compares the task and associated callback's version and transition count to detect staleness.
+func (l *taskExecutor) validateCallbackTask(ms workflow.MutableState, task CallbackAddressingTask) error {
+	ns := ms.GetNamespaceEntry()
+	callbackID := task.GetCallbackID()
+	callback, ok := ms.GetExecutionInfo().GetCallbacks()[callbackID]
+	if !ok {
+		// Some unexpected process deleted the callback from mutable state or mutable state is stale.
+		l.logger.Debug("callback not found in mutable state", tag.WorkflowNamespaceID(ns.ID().String()), tag.Task(task))
+		return queues.NewTaskStateMismatchError(fmt.Sprintf("invalid callback ID for task: %s", callbackID), true)
+	}
+
+	if err := l.validateTaskVersion(ns, callback.GetNamespaceFailoverVersion(), task); err != nil {
+		return err
+	}
+
+	if callback.TransitionCount < task.GetTransitionCount() {
+		l.logger.Debug("callback transitions < task transitions", tag.WorkflowNamespaceID(ns.ID().String()), tag.Task(task))
+		return queues.NewTaskStateMismatchError("callback transitions < task transitions", true)
+	}
+	if callback.TransitionCount > task.GetTransitionCount() {
+		l.logger.Debug("callback transitions > task transitions", tag.WorkflowNamespaceID(ns.ID().String()), tag.Task(task))
+		return queues.NewTaskStateMismatchError("callback transitions < task transitions", false)
```

Comment:
ok I think there's no conflict here with my PR for checking stale ms using shard clock.
The check here is enough to tell is mutable state is stale or not. Trigging shard unload is not that important. 

My view for those checks are:
1. The framework will responsible for returning an up to date mutable state with version that matches the task. This can be done for all tasks by having a versioned state transition tree (similar to version history but for state transition) in mutable state which is updated for all state transitions. All task generated will have a field storing a node of the tree. For validation, we just check if the node in the task is on the current branch of the tree.
2. Check for stale tasks. I think the assumption here is that all task processing logic should be idempotent, task can be reprocessed or regenerated at any time. Then this check is unnecessary, or a responsibility of the executor logic itself.